crates/proxy-server/src/main.rs

```diff
@@ -1,81 +1,157 @@
+             or use --listen for legacy CLI mode."
+        );
     }
 
     let (_no_file_soft, no_file_hard) = getrlimit(Resource::RLIMIT_NOFILE)?;
     setrlimit(Resource::RLIMIT_NOFILE, no_file_hard, no_file_hard).context("setrlimit NOFILE")?;
 
-    for endpoint in &opts.listen {
-        start_listener(
-            endpoint,
-            std::time::Duration::from_secs(opts.timeout_seconds),
-            opts.no_splice,
-        )
-        .await?;
-    }
+    kumo_server_common::panic::register_panic_hook();
 
-    tokio::signal::ctrl_c().await?;
-
-    Ok(())
+    tokio::runtime::Builder::new_multi_thread()
+        .enable_all()
+        .on_thread_park(kumo_server_memory::purge_thread_cache)
+        .build()
+        .unwrap()
+        .block_on(async move { run(opts).await })
 }
 
-async fn start_listener(
-    endpoint: &str,
-    timeout: std::time::Duration,
-    no_splice: bool,
-) -> anyhow::Result<()> {
-    let listener = TcpListener::bind(endpoint)
+async fn perform_init() -> anyhow::Result<()> {
+    tracing::info!("Version is {}", version_info::kumo_version());
+    let mut config = config::load_config().await?;
+
+    let proxy_init_sig = CallbackSignature::<(), ()>::new("proxy_init");
+
+    config
+        .async_call_callback(&proxy_init_sig, ())
         .await
-        .with_context(|| format!("failed to bind to {endpoint}"))?;
-
-    let addr = listener.local_addr()?;
-    log::info!("proxy listener on {addr:?}");
-
-    tokio::spawn(async move {
-        loop {
-            let (socket, peer_address) = match listener.accept().await {
-                Ok(tuple) => tuple,
-                Err(err) => {
-                    log::error!("accept failed: {err:#}");
-                    return;
-                }
-            };
-
-            tokio::spawn(async move {
-                if let Err(err) =
-                    proxy_handler::handle_proxy_client(socket, peer_address, timeout, no_splice)
-                        .await
-                {
-                    log::error!("proxy session error: {err:#}");
-                }
-            });
-        }
-    });
+        .context("in proxy_init event")?;
+    config.put();
+
     Ok(())
 }
+
+async fn signal_shutdown() {
+    tracing::info!("shutting down");
+}
+
+async fn run(opts: Opt) -> anyhow::Result<()> {
+    kumo_server_runtime::assign_main_runtime(tokio::runtime::Handle::current());
+
+    // Determine the policy file to use
+    let (policy_path, _temp_file) = if opts.is_legacy_mode() {
+        // Legacy mode: generate a temporary Lua config from CLI options
+        let temp_file = opts.generate_legacy_config()?;
+        let path = temp_file.path().to_path_buf();
+        (path, Some(temp_file))
+    } else {
+        // Modern mode: use the provided config file
+        let path = opts
+            .proxy_config
+            .unwrap_or_else(|| PathBuf::from("/opt/kumomta/etc/proxy/init.lua"));
+        (path, None)
+    };
```

Comment:
It's slightly more idiomatic and cleaner to flip this around this way; we can remove the impossible `unwrap_or_else` case and its fallback to a default path that we'll never use:

```suggestion
    let legacy_file;
    
    let policy_path = match opts.proxy_config {
        Some(path) => path,
        None => {
           // Legacy mode: generate a temporary Lua config from CLI options
           legacy_file = opts.generate_legacy_config()?;
           temp_file.path().to_path_buf()
        }
    };
```

That `let legacy_file;` line looks weird on its own up there at first, but it is the idiomatic way to hold ownership of the temporary file object and avoid adding boilerplate/noise to thread an optional extra thing around through the match.